.github/workflows/on-push.yaml

```diff
@@ -0,0 +1,254 @@
+name: On Push
+on: [push]
+
+concurrency:
+  # limit concurrency of entire workflow runs for a specific branch
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+env:
+  NODE_VERSION: 22.1.0
+  BUF_VERSION: 1.42.0
+  PYTHON_VERSION: 3.13
+  CARGO_INCREMENTAL: 0
+  CARGO_PROFILE_DEV_STRIP: "debuginfo"
+
+jobs:
+
+  # Runs various lints and checks for the project, including Rustfmt, Clippy,
+  # Protobuf lints, and pnpm lints.
+  lint:
+    name: Run Lints
+    runs-on: ubuntu-24.04
+    steps:
+      - uses: rui314/setup-mold@v1
+        with:
+          make-default: true
+      - uses: actions/checkout@v4
+        with:
+          submodules: "recursive"
+      - uses: pnpm/action-setup@v4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+          cache: "pnpm"
+      - uses: arduino/setup-protoc@v3
+        with:
+          version: "25.3"
+      - uses: bufbuild/buf-action@v1
+        with:
+          version: ${{ env.BUF_VERSION }}
+          setup_only: true
+      - uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          components: clippy, rustfmt
+          cache-key: "rust-tests"
+      - name: Install Package Dependencies
+        run: make install
+      - name: Lint (Rustfmt)
+        run: cargo fmt --all -- --check
+      - name: Lint (Clippy)
+        run: cargo clippy -- -D warnings
+      - name: Lint (pnpm)
+        run: pnpm --recursive run lint
+      - name: Lint (Protobuf)
+        run: |
+          buf format --diff --exit-code
+          buf lint
+        working-directory: ./protobufs
+      - name: Typecheck (pnpm)
+        run: pnpm --recursive typecheck
+
+  # Builds the Rust test artifacts for the project, packages them as Nextest
+  # archives and uploads them as artifacts. This job is used as a dependency for
+  # the `unit-tests` and `integration-tests` jobs.
+  build-tests:
+    name: Build Test Artifacts
+    runs-on: ubuntu-24.04
+    steps:
+      - uses: rui314/setup-mold@v1
+        with:
+          make-default: true
+      - uses: actions/checkout@v4
+        with:
+          submodules: "recursive"
+      - uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          cache-key: "rust-tests"
+      # - uses: pnpm/action-setup@v4
+      # - uses: actions/setup-node@v3
+      #   with:
+      #     node-version: ${{ env.NODE_VERSION }}
+      #     cache: "pnpm"
+      # - uses: arduino/setup-protoc@v3
+      #   with:
+      #     version: "25.3"
+      - uses: taiki-e/install-action@v2
+        with:
+          tool: nextest@0.9.82
+      # - name: Install Package Dependencies
+      #   run: make install
+      - name: Build Tests
+        run: make test-build
+      - name: Create Nextest Archives
+        run: make nextest-archive
```

Comment:
Why is this necessary?

Reply:
Oh I see, this compiles the tests and stuff, or something like that.

Reply:
Yup -- we do one single build of all of the test artifacts and then package them into two archives: one for tests that can be run in parallel (unit tests right now), and one for tests that must be run serially.

The running of the tests then no longer need rust, only the nextest binary. So we use this to ship the artifacts to the jobs which actually run the tests without needing to build them again.